Retain hooks when a filtered helm test is interrupted

lheinlen opened the pull request:
Signed-off-by: Lucas Heinlen <lheinlen@onecause.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/master/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
This PR fixes #9398
**Special notes for your reviewer**:
My goal was to not change any functionality other than to retain all of the hooks in the stored release object through the entire `helm test` process. Hopefully, I was successful in that. :laughing: 

**If applicable**:
- [x] this PR contains documentation - No change in functionality
- [x] this PR contains unit tests - I believe that this should be covered by existing tests
- [x] this PR has been tested for backwards compatibility - No change in functionality

Commit:
Retain hooks when a filtered helm test is interrupted

Signed-off-by: Lucas Heinlen <lheinlen@onecause.com>

bacongobbler commented:
I'm confused how `ExecutionDisabled` is set when the `--filter` flag is used. Right now it seems that every hook has this set to `false`.

How is the CLI intended to use this flag? If it's meant to be unused, then why should a boolean be set at all?

lheinlen commented:
`ExecutionDisabled` is being set in the filter-related logic of the `ReleaseTesting` action defined in `pkg/action/release_testing.go` which is the action executed when `helm test` is run. It will currently only be set by `helm test` when the `--filter` flag is used.

bacongobbler commented:
:headsmack: Sorry. I confused `*_testing.go` with `_test.go`, which is the convention used for Go tests. I see what you're saying now. Thanks for talking me through it.

Would you mind writing a unit test to ensure future regressions do not re-introduce this bug?

bacongobbler commented on pkg/release/hook.go:
```diff
@@ -76,6 +76,8 @@ type Hook struct {
 	// DeletePolicies are the policies that indicate when to delete the hook
 	DeletePolicies []HookDeletePolicy `json:"delete_policies,omitempty"`
+	// ExecutionDisabled is set to true when a hook has been filtered out and should not be executed
+	ExecutionDisabled bool `json:"-"`
```
Given that this boolean is controlled by the filters, it's probably best to just make this a private variable instead. If I'm reading the code correctly, attempting to create a `release.Hook{ExecutionDisabled: true}` vs a `release.Hook{}` will not change its behaviour. In other words it's an internal state machine: it should not be exposed to the public API at this time.

lheinlen replied:
Since the action and this structure and in different packages, if I make it a private field, it will be inaccessible from the action. I can certainly do this, but I would need to make a public setter like `DisableExecution` or something similar to toggle the field. If that would be preferable, I am happy to make the change.

Another option would be to alter `execHook` rather than the hook structure. I briefly considered doing that but it initially seemed cleaner to mark the hook itself. Perhaps I was mistaken about that! I don't want to muddy the public API. I could do something like pass in a `activeHooks` variable to `execHook` and, if it is `nil`, default it to `rl.Hooks`. Then pass in the filtered hooks from `helm test` and `nil` from all of the other call locations.

lheinlen commented:
> Would you mind writing a unit test to ensure future changes do not re-introduce this bug?

Sure. I'll take a look at writing a few tests for the `ReleaseTesting` action. I don't think there are any currently (or I've just missed them).

Commit:
Add basic tests for ReleaseTesting action

Signed-off-by: Lucas Heinlen <lheinlen@onecause.com>

lheinlen commented:
The tests that I added are somewhat simplistic but they...
- Verify that the correct hooks were executed with a few different filter setups
- Verify that hooks are not lost when the process stops abruptly after the release is updated
I rebased to put the tests first and ran them against the previous implementation. The "crash" test failed as expected and succeeded after this change.

Commit:
Revert previous implementation

Signed-off-by: Lucas Heinlen <lheinlen@onecause.com>

Commit:
Retain hooks when a filtered helm test is interrupted

Signed-off-by: Lucas Heinlen <lheinlen@onecause.com>

lheinlen commented:
I reworked this change to not touch the `release.Hook` struct. I left the previous commit in the history for comparison.

lheinlen commented:
@bacongobbler Is there anything else I could do to improve this PR to help it along? We've run into this a few more times recently.

lheinlen commented:
@bacongobbler Sorry to bother but have you had a chance to look over these changes? We continue to run into this issue occasionally. I would be happy to rework this in another way if it's not satisfactory.

bacongobbler commented:
Sorry been busy with other work. Probably won't be able to get to this for a while. I would suggest reaching out to the other core maintainers and see if they have time to review this.

Thanks for doing the work!

bacongobbler commented:
As per our contributing policy this will need sign-off from two maintainers, FYI.

https://github.com/helm/helm/blob/master/CONTRIBUTING.md#size-labels

kashook commented:
Per the recommendation of @bacongobbler, reaching out to some other helm maintainers (just picked randomly from [this list](https://github.com/helm/helm/blob/main/OWNERS)) to see if anyone can review this PR.

@mattfarina, @adamreese, @technosophos, I know people are busy, but is there any chance you can review this bug fix PR at some point? The actual fix is pretty small. @lheinlen added unit tests to this area of code that previously didn't have any (which made the PR larger). The tests seem to prove that the fix works and didn't break anything.

krmichelos commented:
@hickeyma, @jdolitsky, @marckhouzam, @prydonius, @SlickNik could a couple of you find some time to take a look at this PR?  A pretty good chunk of the changes are tests and it would be very helpful to get this bug fixed that is now over four months old.

lheinlen commented:
We just ran into this again. We have a process which runs many helm tests in parallel and it failed in such a way that multiple of our releases lost their hooks. Due to the nature of this issue, our only choice was to re-deploy them all.

I would love to get this fixed. If a different solution is preferred, I would be more than happy to rework the code.

krmichelos commented:
@bacongobbler, @mattfarina, @adamreese, @technosophos, @hickeyma, @jdolitsky, @marckhouzam, @prydonius, @SlickNik this is now about 6 months old and not that big of a PR given that most of the lines are tests.

mateinavidad commented:
Any chance of merging this anytime soon? It's currently blocking us from switching to using `helm test --filter` in production.

Commit:
Merge branch 'main' into retain-hooks-on-test-interrupt

lheinlen commented:
I've merged main and resolved the conflicts. If there is anything else I can do to help move this along, please let me know.

mateinavidad commented:
Any chance of reviewing this and merging it? @bacongobbler, @mattfarina, @adamreese, @technosophos, @hickeyma, @jdolitsky, @marckhouzam, @prydonius, @SlickNik

Commit:
Fix spelling error

Signed-off-by: Lucas Heinlen <lheinlen@onecause.com>

code-bringer commented:
Actually, this feature will help us to run helm tests in parallel, which is great, So it would be nice if maintainers will have time to review it.

briceo commented:
Just tagging the maintainers of this repo (pulled from [this list](https://github.com/helm/helm/blob/main/OWNERS))
@hickeyma @joejulian @jdolitsky @marckhouzam @mattfarina @sabre1041 @scottrigby @technosophos
Hello, is there any chance that this could get merged?  We keep running into this issue, and it would be great if we can get it resolved.  Is there something else that needs to be completed before it can be merged?  Thanks!

scottrigby commented:
> Just tagging the maintainers of this repo (pulled from [this list](https://github.com/helm/helm/blob/main/OWNERS)) @hickeyma @joejulian @jdolitsky @marckhouzam @mattfarina @sabre1041 @scottrigby @technosophos Hello, is there any chance that this could get merged? We keep running into this issue, and it would be great if we can get it resolved. Is there something else that needs to be completed before it can be merged? Thanks!

Hi @briceo 👋 we're doing our best to work through reviewing the open PRs. We now have a bigger team of folks helping us prepare the PRs for review, so expecting this to go faster moving ahead. Thanks for your patience. @lheinlen we appreciate the work you put into this PR, please keep contributing 🙂
